components/browser/session/src/main/java/mozilla/components/browser/session/SessionManager.kt

```diff
@@ -16,17 +18,50 @@ import mozilla.components.support.base.observer.ObserverRegistry
 @Suppress("TooManyFunctions")
 class SessionManager(
     val engine: Engine,
+    val initialSessions: SessionsSnapshot? = null,
+    val defaultSession: (() -> Session)? = null,
     delegate: Observable<SessionManager.Observer> = ObserverRegistry()
 ) : Observable<SessionManager.Observer> by delegate {
     private val values = mutableListOf<Session>()
+    // Access must be guarded by synchronizing on `values`.
     private var selectedIndex: Int = NO_SELECTION
 
+    init {
+        initialSessions?.forEach {
+            add(it.session, selected = it.selected, engineSession = it.engineSession)
+        }
+
+        if (defaultSession != null && size == 0) {
+            add(defaultSession.invoke())
+        }
+    }
+
     /**
      * Returns the number of session including CustomTab sessions.
      */
     val size: Int
         get() = synchronized(values) { values.size }
 
+    /**
+     * Returns a snapshot suitable for persisting for "session restore" purposes.
+     */
+    val snapshot: SessionsSnapshot
+        get() = synchronized(values) {
+            // Filter out CustomTab and private sessions.
+            // We're using 'values' directly instead of 'sessions' to get benefits of a sequence.
+            values.asSequence()
```

Comment:
I've used `values` directly for the sake of some extra perf - at least that's what `sequence` claims to provide for multi-step collection manipulations. https://blog.kotlin-academy.com/effective-kotlin-use-sequence-for-bigger-collections-with-more-than-one-processing-step-649a15bb4bf

Arguably this might also be nice for other reasons - it's explicit, so you don't need to go hunting around the class while reading this code, and changes to `sessions` won't have unintended consequences on session restore.

Reply:
OK, that makes sense. It was just a nit. Marking as "resolved" :)